aip/0160.md

```diff
@@ -0,0 +1,167 @@
+
+**Note:** Google provides implementations for the filtering language in both
+[C++][cel-cpp] and [Go][cel-go] as a discrete component of [CEL][]. The use of
+these implementations is encouraged.
+
+### Expressions
+
+Expressions are literal values in the filtering syntax ("42", "Victor Hugo").
+Expressions can be pieced together with operators to make other expressions.
+
+### Logical Operators
+
+Filtering implementations **must** provide the binary operators `AND` and `OR`,
+as well as a whitespace operator:
+
+| Operator | Example       | Meaning                                |
+| -------- | ------------- | -------------------------------------- |
+| `AND`    | `a AND b`     | True if `a` and `b` are true.          |
+| `OR`     | `a OR b OR c` | True if any of `a`, `b`, `c` are true. |
+
+The whitespace operator (example: `a b c`) represents a fuzzy variant of `AND`,
+which is suitable for result ranking and term rewrite.
+
+**Note:** To match common patterns of speech, the `OR` operator has higher
+precedence than `AND`, unlike what is found in most programming languages. The
+expression `a AND b OR c` evaluates: `a AND (b OR c)`. API documentation and
+examples **should** encourage the use of explicit parentheses to avoid
+confusion.
+
+### Negation Operators
+
+Filtering implementations **must** provide the unary operators `NOT` and `-`.
+These are used interchangeably.
+
+| Operator | Example | Meaning                  |
+| -------- | ------- | ------------------------ |
+| `NOT`    | `NOT a` | True if `a` is not true. |
+| `-`      | `-a`    | True if `a` is not true. |
+
+### Comparison Operators
+
+Filtering implementations **must** provide the binary comparison operators `=`,
+`!=`, `<`, `>`, `<=`, and `>=`.
+
+| Operator | Example      | Meaning                                         |
+| -------- | ------------ | ----------------------------------------------- |
+| `=`      | `a = true`   | True if `a` is true.                            |
+| `!=`     | `a != 42`    | True unless `a` equals 42.                      |
+| `<`      | `a < 42`     | True if `a` is a numeric value below 42.        |
+| `>`      | `a > "foo"`  | True if `a` is lexically ordered after "foo".   |
+| `<=`     | `a <= "foo"` | True if `a` is "foo" or lexically before it.    |
+| `>=`     | `a >= 42`    | True if `a` is a numeric value of 42 or higher. |
+
+Because filters are accepted as querty strings, type conversion takes place to
+translate the string to the appropriate strongly-typed value:
+
+- Enums expect the enum's string representation (case-sensitive).
+- Booleans expect `true` and `false` literal values.
+- Numbers expect the standard integer or float representations. For floats,
+  exponents are supported (e.g. `2.997e9`).
+- [Durations][] expect a numeric representation followed by an `s` suffix (for
+  seconds). Examples: `20s`, `1.2s`.
+- [Timestamps][] expect an [RFC-3339][] formatted string (e.g.
```

Comment:
Timezones => UTC offsets
(If we must use the term, let's make it "time zones" rather than "timezones" - but what's being specified here really isn't a time zone.)

Reply:
Thanks, "UTC offsets" strike me as almost certainly the correct term. (@TristonianJones, I assume you do not accept "EST", "EDT", etc., correct?)

Reply:
The date time values are expected to be in the `RFC-3339` format which optionally supports the use of `-+HH:MM` offsets from UTC within the string. Does that make sense?

datetime in UTC: `2012-04-21T11:30:00Z`
datetime with TZ: `2012-04-21T11:30:00-04:00`

Reply:
Okay, so "UTC offsets" sounds most precise.

Reply:
And I fixed my example (to add the colon).